flocker/node/agents/cinder.py

```diff
@@ -169,36 +180,60 @@ def wait_for_volume(volume_manager, expected_volume,
             )
 
 
+def _extract_nova_server_addresses(addresses):
+    """
+    :param dict addresses: A ``dict`` mapping OpenStack network names
+        to lists of address dictionaries in that network assigned to a
+        server.
+    :return: A ``set`` of all the IPv4 and IPv6 addresses from the
+        ``addresses`` attribute of a ``Server``.
+    """
+    all_addresses = set()
+    for network_name, addresses in addresses.items():
+        for address_info in addresses:
+            all_addresses.add(address_info['addr'])
+
+    return all_addresses
+
+
 @implementer(IBlockDeviceAPI)
 class CinderBlockDeviceAPI(object):
     """
     A cinder implementation of ``IBlockDeviceAPI`` which creates block devices
     in an OpenStack cluster using Cinder APIs.
     """
-    def __init__(self, cinder_volume_manager, nova_volume_manager, cluster_id):
+    def __init__(self,
+                 cinder_volume_manager,
+                 nova_volume_manager, nova_server_manager,
+                 cluster_id):
         """
         :param ICinderVolumeManager cinder_volume_manager: A client for
             interacting with Cinder API.
         :param INovaVolumeManager nova_volume_manager: A client for interacting
             with Nova volume API.
+        :param INovaServerManager nova_server_manager: A client for interacting
+            with Nova servers API.
         :param UUID cluster_id: An ID that will be included in the names of
             Cinder block devices in order to associate them with a particular
             Flocker cluster.
         """
         self.cinder_volume_manager = cinder_volume_manager
         self.nova_volume_manager = nova_volume_manager
+        self.nova_server_manager = nova_server_manager
         self.cluster_id = cluster_id
 
     def compute_instance_id(self):
         """
-        Look up the Xen instance ID for this node.
+        Find the Nova API server with the same non-loopback IP
+        addresses as this node.
+        That server is assumed to be this node.
+        Return the instance ID of that server.
```

Comment:
Convert this to proper markup.  Also, what if no match is found?  The error case is probably worth handling explicitly.  We may also want to log enough information here that we could debug a failure caused by someone having a network configuration we didn't anticipate.